### PR TITLE
311: Catch divergence errors

### LINKED ISSUE #311: Certain errors from the function or optimizer evaluation should be caught
Examples:

- `L-BFGS-B needs finite values of 'fn'`
- `NA/NaN Hessian evaluation`

### PR DESCRIPTION
**Pull request by danielinteractive**
closes #311

### CODE REVIEW

File: .Rbuildignore [hunk 1]
```diff
@@ -17,3 +17,4 @@ simulations
 ^doc$
 ^Meta$
 ^data-raw$
+cache
```
Comment (danielinteractive) [inline .Rbuildignore:20]:
```suggestion
^cache$
```

File: R/fit.R [hunk 2]
```diff
@@ -415,12 +418,21 @@ mmrm <- function(formula,
...
+      formula_parts = formula_parts,
+      control = control
+    )
+    if (length(fit) == 0) {
+      warning(paste0(
+        "Divergence with optimizer ", names(control$optimizers[1L]), " due to problems: ",
+        toString(attr(fit, "divergence"))
+      ))
+    }
+    control$optimizers <- control$optimizers[-1]
```
Comment (clarkliming) [inline R/fit.R:434]: if all optimizer fails, `fit` is length 0 and there are no successful optimizers. then is there need to call `refit_multiple_optimizers` again?

File: R/fit.R [hunk 3]
```diff
@@ -40,6 +40,16 @@ fit_single_optimizer <- function(formula,
...
+  to_remove <- list(
+    # Transient visit to invalid parameters.
+    warnings = c("NA/NaN function evaluation")
+  )
+  as_diverged <- list(
+    errors = c(
+      "NA/NaN Hessian evaluation",
+      "L-BFGS-B needs finite values of 'fn'"
+    )
+  )
```
Comment (clarkliming) [inline R/fit.R:52]: if this is something static, should it be defined somewhere not inside the function but in the package environment?

Comment (danielinteractive) [reply R/fit.R:52]: I think it is ok, we had it like that before too for the warnings to remove - it is not used by any other functions e.g.

File: R/fit.R (hunk 2 above)
Comment (danielinteractive) [reply R/fit.R:434]: this first while part is only finding the first optimizer returning a non-empty fit. Exactly, if all fail then below we will get the error to try multiple optimizers and refit_multiple_optimizers is not called. also have a look at the unit tests I added for these edge cases

Comment (clarkliming) [inline R/fit.R:434]: so this part we modified the `control$optimizers`, and in line 451 we still need the names of all these optmizers to report the error. currently it says

```
Chosen optimizer '' led to problems during model fit:
```

File: R/fit.R [hunk 4]
```diff
@@ -415,12 +418,21 @@ mmrm <- function(formula,
...
-
+  fit <- list()
+  while ((length(fit) == 0) && length(control$optimizers) > 0) {
+    fit <- fit_single_optimizer(
+      tmb_data = tmb_data,
+      formula_parts = formula_parts,
+      control = control
+    )
+    if (length(fit) == 0) {
+      warning(paste0(
```
Comment (clarkliming) [inline R/fit.R:429]: I wonder if we really need this warning (because other divergent issues might be captured by the optimizers, like different convergence code; and other non-converged issue happens we just silently go to other optimizers).  Maybe also good to capture all the warning/errors in `do.call` part of the optimizer so that anyway we have a returned fit in `fit_mmrm` but all the errors/logs are stored. (`tmb_opt` result can be manually constructed with `list(value = NA, par = initial values etc, convergence = 1 etc, error = caught error)`

Comment (danielinteractive) [reply R/fit.R:430]: Good ideas! I thought about this a bit more after our stand up.

- I think I would not move the capturing of errors and warnings further inside, because `fit_mmrm` (one level down) is supposed to be low level, i.e. not capture these. (and if we would consider this, it would be quite big change with separate issue/PR)
- So far all the examples we have seen recently about these divergence errors/warnings were actually pointing towards overparametrized covariance models (instead of just unfortunate optimizer behavior). Therefore I think the warning makes sense to alert the user, while still trying other optimizers.

**Review by clarkliming — APPROVED** [review]
looks good to me!
